File: apfelschlacht/player.py
"""Player and MovingObject classes for Apfelschlacht."""

import random
from collections import deque

import arcade

from .constants import (
    Algorithm,
    SPRITE_SCALING_PLAYER,
    SPRITE_FRAME_COUNT,
    MIN_SPEED,
    MAX_SPEED,
    VALUES_FOR_AVERAGE,
    DEFAULT_SPEED_APPLE,
)

TEXTURE_PATHS = {
    "boy": ":resources:images/animated_characters/female_person/femalePerson_walk{}.png",
    "man": ":resources:images/animated_characters/male_person/malePerson_walk{}.png",
}


class MovingObject:
    """A sprite moving toward a destination at a fixed speed."""

    def __init__(self, dest_x, dest_y, speed, sprite_obj):
        self.sprite_obj = sprite_obj
        dist = arcade.get_distance(
            sprite_obj.center_x, sprite_obj.center_y, dest_x, dest_y
        )
        self.steps = int(dist / speed)
        self.delta_x = (dest_x - sprite_obj.center_x) / self.steps
        self.delta_y = (dest_y - sprite_obj.center_y) / self.steps

    def move(self):
        if self.steps > 0:
            self.steps -= 1
            self.sprite_obj.center_x += self.delta_x
            self.sprite_obj.center_y += self.delta_y
            return True
        return False


class Player:
    """A player character that collects and throws apples."""

    def __init__(self, name, position, screen_dimensions):
        self.name = name
        self.fly_apples = 0
        self.speed_apple = DEFAULT_SPEED_APPLE
        self.fly_apple_numbers = deque()
        self.avg_dist = deque()
        self.screen_width = screen_dimensions[0]
        self.screen_height = screen_dimensions[1]
        self.speed = 6 if name == "boy" else 3
        self.steps = 0
        self.delta_x = 0
        self.delta_y = 0
        self.my_apples = arcade.SpriteList()
        self.next_apple = None
        self.objs_in_motion = []
        self.sprite_frame = 0
        self.sprite_time = 0
        self.combustion_time = 0
        self.texture_offset = 0

        self.sprite = arcade.Sprite()
        self.sprite.center_x = position[0]
        self.sprite.center_y = position[1]

        template = TEXTURE_PATHS[name]
        textures = [
            arcade.load_texture(template.format(i))
            for i in range(SPRITE_FRAME_COUNT)
        ]
        # Duplicate for facing direction (original behavior: indices 8-15 mirror 0-7)
        self.sprite.textures = textures + textures
        self.sprite.texture = self.sprite.textures[0]
        self.sprite.scale = SPRITE_SCALING_PLAYER

    def change_speed(self, change):
        self.speed = max(MIN_SPEED, min(MAX_SPEED, self.speed + change))

    def find_next_apple(self, algorithm, stats_recorder=None):
        nearest_apple = None
        dist = float("inf")
        x = self.sprite.center_x
        y = self.sprite.center_y

        if algorithm == Algorithm.NEAREST_APPLE:
            for apple in self.my_apples:
                a_dist = arcade.get_distance(x, y, apple.center_x, apple.center_y)
                if a_dist < dist:
                    dist = a_dist
                    nearest_apple = apple

        elif algorithm == Algorithm.RANDOM:
            if len(self.my_apples) > 0:
                nearest_apple = self.my_apples[random.randrange(len(self.my_apples))]
                dist = arcade.get_distance(
                    x, y, nearest_apple.center_x, nearest_apple.center_y
                )

        if algorithm != Algorithm.SPONTANEOUS_COMBUSTION:
            if nearest_apple is not None:
                self.avg_dist.append(dist)
                if len(self.avg_dist) > VALUES_FOR_AVERAGE:
                    self.avg_dist.popleft()
                    if stats_recorder:
                        stats_recorder.record_distance(self.name, dist)
                self.steps = int(dist / self.speed) or 1
                self.delta_x = (nearest_apple.center_x - x) / self.steps
                self.delta_y = (nearest_apple.center_y - y) / self.steps
                self.next_apple = nearest_apple
        else:
            self._spontaneous_combustion()

    def _spontaneous_combustion(self):
        combusting_apples = 0
        if self.combustion_time > (1 / self.speed):
            if len(self.my_apples) > 0:
                combusting_apples = round(
                    random.randrange(len(self.my_apples))
                    * (0.05 + self.speed / 100)
                )
                self.combustion_time = 0

        while len(self.my_apples) > 0 and combusting_apples > 0:
            apple = self.my_apples[round(random.randrange(combusting_apples))]
            combusting_apples -= 1
            offset = self.screen_width / 2 + 20 if self.name == "man" else 0
            dest_x = random.randrange(int(self.screen_width / 2 - 20)) + offset + 10
            dest_y = random.randrange(self.screen_height - 220) + 10
            fly_apple = MovingObject(dest_x, dest_y, self.speed_apple, apple)
            self.fly_apples += 1
            self.objs_in_motion.append(fly_apple)
            self.my_apples.remove(apple)

File: web/src/lib/engine/Player.ts
import { Algorithm, MIN_SPEED, MAX_SPEED, VALUES_FOR_AVERAGE, DEFAULT_SPEED_APPLE, SPRITE_FRAME_COUNT } from './constants';
import { MovingObject } from './MovingObject';
import type { Apple } from './types';
import type { StatsRecorder } from './StatsRecorder';

export class Player {
	name: string;
	x: number;
	y: number;
	speed: number;
	speedApple: number;
	flyApples: number;
	screenWidth: number;
	screenHeight: number;

	myApples: Apple[] = [];
	nextApple: Apple | null = null;
	objsInMotion: MovingObject[] = [];

	steps = 0;
	deltaX = 0;
	deltaY = 0;

	spriteFrame = 0;
	spriteTime = 0;
	textureOffset = 0;

	halfLife = 10;
	flyAppleNumbers: number[] = [];
	avgDist: number[] = [];

	constructor(name: string, x: number, y: number, speed: number, screenWidth: number, screenHeight: number) {
		this.name = name;
		this.x = x;
		this.y = y;
		this.speed = speed;
		this.speedApple = DEFAULT_SPEED_APPLE;
		this.flyApples = 0;
		this.screenWidth = screenWidth;
		this.screenHeight = screenHeight;
	}

	changeSpeed(delta: number): void {
		this.speed = Math.max(MIN_SPEED, Math.min(MAX_SPEED, this.speed + delta));
	}

	findNextApple(algorithm: Algorithm, dt: number, stats?: StatsRecorder): void {
		let nearestApple: Apple | null = null;
		let dist = Infinity;

		if (algorithm === Algorithm.NearestApple) {
			for (const apple of this.myApples) {
				const dx = this.x - apple.x;
				const dy = this.y - apple.y;
				const d = Math.sqrt(dx * dx + dy * dy);
				if (d < dist) {
					dist = d;
					nearestApple = apple;
				}
			}
		} else if (algorithm === Algorithm.Random) {
			if (this.myApples.length > 0) {
				nearestApple = this.myApples[Math.floor(Math.random() * this.myApples.length)];
				const dx = this.x - nearestApple.x;
				const dy = this.y - nearestApple.y;
				dist = Math.sqrt(dx * dx + dy * dy);
			}
		}

		if (algorithm !== Algorithm.SpontaneousCombustion) {
			if (nearestApple !== null) {
				this.avgDist.push(dist);
				if (this.avgDist.length > VALUES_FOR_AVERAGE) {
					this.avgDist.shift();
					if (stats) {
						stats.recordDistance(this.name, dist);
					}
				}
				this.steps = Math.max(1, Math.floor(dist / this.speed));
				this.deltaX = (nearestApple.x - this.x) / this.steps;
				this.deltaY = (nearestApple.y - this.y) / this.steps;
				this.nextApple = nearestApple;
			}
		} else {
			this.spontaneousCombustion(dt);
		}
	}

	private spontaneousCombustion(dt: number): void {
		if (this.halfLife <= 0 || this.myApples.length === 0) return;
		const p = 1 - Math.exp(-Math.LN2 / this.halfLife * dt);

		for (let i = this.myApples.length - 1; i >= 0; i--) {
			if (Math.random() < p) {
				const apple = this.myApples[i];
				const offset = this.name === 'boy' ? this.screenWidth / 2 + 20 : 0;
				const destX = Math.floor(Math.random() * (this.screenWidth / 2 - 20)) + offset + 10;
				const destY = Math.floor(Math.random() * (this.screenHeight - 20)) + 10;
				const flyApple = new MovingObject(destX, destY, this.speedApple, apple);
				this.flyApples++;
				this.objsInMotion.push(flyApple);
				this.myApples.splice(i, 1);
			}
		}
	}

	getAvgDist(): number {
		if (this.avgDist.length < VALUES_FOR_AVERAGE) return 0;
		return this.avgDist.reduce((a, b) => a + b, 0) / this.avgDist.length;
	}

	getAvgFlyApples(): number {
		if (this.flyAppleNumbers.length < VALUES_FOR_AVERAGE) return 0;
		return this.flyAppleNumbers.reduce((a, b) => a + b, 0) / this.flyAppleNumbers.length;
	}

	updateAnimation(dt: number): void {
		this.spriteTime += dt;
		if (this.spriteTime > 0.1) {
			this.spriteTime = 0;
			this.spriteFrame = (this.spriteFrame + 1) % SPRITE_FRAME_COUNT;
			this.textureOffset = this.deltaX < 0 ? SPRITE_FRAME_COUNT : 0;
		}
	}
}
